File: Data Structures & Algorithms/meeting-schedule-ii/submission-6.py
"""
Definition of Interval:
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class Solution:
    def minMeetingRooms(self, intervals: List[Interval]) -> int:
        if not intervals:
            return 0
        
        count = 0
        s = [interval.start for interval in intervals]
        e = [interval.end for interval in intervals]
        s.sort()
        e.sort()
        res_count = 0
        i, j = 0, 0
        while i < len(s) and j < len(e):
            if i < len(s):
                if s[i] < e[j]:
                    count += 1
                    i += 1
                else: 
                    count -= 1
                    j += 1
                res_count = max(count, res_count)

        return res_count

File: Data Structures & Algorithms/insert-new-interval/submission-6.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> result = new ArrayList<>();
        for (int[] interval : intervals){
            /*
            Noted that the given intervals is sorted in ascending order

            Compare the current interval end and new interval start time.
            If the new interval start time is smaller or equal than the current interval end time, 
            it means these two interval is overlapped, thus these two interval will be merge.

            */
            if (interval[1] < newInterval[0]){
                result.add(interval);
            } else if (interval[0] > newInterval[1]){
                result.add(newInterval);
                newInterval = interval;
            }else{
                newInterval[0] = Math.min(interval[0], newInterval[0]);
                newInterval[1] = Math.max(interval[1], newInterval[1]);
            }
        }
        result.add(newInterval);
        return result.toArray(new int[result.size()][]);
    }
}